plonky3/src/prover/mod.rs

```diff
@@ -92,18 +114,52 @@ mod tests {
         let witness_callback = pipeline.witgen_callback().unwrap();
         let witness = pipeline.compute_witness().unwrap();
 
-        let proof = Plonky3Prover::new(&pil).prove(&witness, witness_callback);
+        let proof = Plonky3Prover::new(&pil).prove(&witness, witness_callback, None);
+
+        assert!(proof.is_ok());
+    }
+
+    fn run_test_goldilocks_publics(pil:& str,  publics: Vec<Goldilocks>) {
+        let mut pipeline = Pipeline::<GoldilocksField>::default().from_pil_string(pil.to_string());
+
+        let pil = pipeline.compute_optimized_pil().unwrap();
+        let witness_callback = pipeline.witgen_callback().unwrap();
+        let witness = pipeline.compute_witness().unwrap();
+
+        let proof = Plonky3Prover::new(&pil).prove(&witness, witness_callback, Some(publics));
 
         assert!(proof.is_ok());
     }
 
     #[test]
-    #[should_panic = "not implemented"]
+    // #[should_panic = "not implemented"]
     fn publics() {
         let content = "namespace Global(8); pol witness x; x * (x - 1) = 0; public out = x(7);";
         run_test_goldilocks(content);
     }
 
+    #[test]
+    #[should_panic = "called `Result::unwrap()` on an `Err` value: OodEvaluationMismatch"]
+    fn public_inputs() {
+        let content = r#"
+        namespace Add(8);
+            col witness x;
+            col witness y;
+            col witness z;
+            x + y = z;
+
+            public outx = x(7);
+            public outy = y(7);
+            public outz = z(7);
+        "#;
+        let publics = vec![
+            cast_to_goldilocks(GoldilocksField::from(0)),
+            cast_to_goldilocks(GoldilocksField::from(1)),
+            cast_to_goldilocks(GoldilocksField::from(1)),
+            ];
+        run_test_goldilocks_publics(content, publics)
```

Comment:
then here we can `assert_eq!(run_test_goldilocks(content, malicious_public_inputs), Err("verification failed"))` instead of `should_panic`